File: prova-senior-java/src/main/java/org/aurum/Calculator.java
package org.aurum;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

public class Calculator {

	public static void main(String[] args) {
		System.out.println(new Calculator().calculateProportionalDifferencePrice(200, 100, "2025-04-24", "MONTHLY"));
	}



	public double calculateProportionalDifferencePrice(
			double oldPrice,
			double newPrice,
			String dueDate,
			String recurrencyType
	) {
		LocalDate dueDateFormat = LocalDate.parse(dueDate);
		int monthsOfRecurrence;

		switch (recurrencyType) {
			case "MONTHLY":
				monthsOfRecurrence = 1;
				break;
			case "SEMIANNUAL":
				monthsOfRecurrence = 6;
				break;
			case "YEARLY":
			case "YEARLY_SPLITED":
				monthsOfRecurrence = 12;
				break;
			default:
				throw new IllegalArgumentException("Unknown recurrence type");
		}

		LocalDate startDateFormat = dueDateFormat.minusMonths(monthsOfRecurrence);
		LocalDate today = LocalDate.now();
		long daysLeft = ChronoUnit.DAYS.between(startDateFormat, today);

		if (daysLeft > 0) {
			long totalPlanDays = ChronoUnit.DAYS.between(startDateFormat, dueDateFormat);
			long remainingDays = totalPlanDays - daysLeft;
			double totalAmountRemainingPayed = 0.0;
			if (remainingDays > 0) {
				double currentPriceByDay = oldPrice / totalPlanDays;
				totalAmountRemainingPayed = currentPriceByDay * remainingDays;
			}
			double newPriceByDay = newPrice / totalPlanDays;
			double totalAmountRemainingNotPayed = newPriceByDay * remainingDays;
			return totalAmountRemainingNotPayed - totalAmountRemainingPayed;
		} else {
			return newPrice - oldPrice;
		}
	}
}

File: prova-senior-kotlin/src/main/java/org/aurum/Calculator.kt
package org.aurum

import java.time.LocalDate
import java.time.temporal.ChronoUnit

class Calculator {

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            println(Calculator().calculateProportionalDifferencePrice(100.0, 300.0, "2025-04-16", "MONTHLY"))
        }
    }

    fun calculateProportionalDifferencePrice(
        oldPrice: Double,
        newPrice: Double,
        dueDate: String,
        recurrencyType: String
    ): Double {
        val dueDateFormat = LocalDate.parse(dueDate)
        val monthsOfRecurrence = when (recurrencyType) {
            "MONTHLY" -> 1
            "SEMIANNUAL" -> 6
            "YEARLY", "YEARLY_SPLITED" -> 12
            else -> throw IllegalArgumentException("Unknown recurrence type")
        }
        val startDateFormat = dueDateFormat.minusMonths(monthsOfRecurrence.toLong())
        val today = LocalDate.now()
        val daysLeft = ChronoUnit.DAYS.between(startDateFormat, today)

        return if (daysLeft > 0) {
            val totalPlanDays = ChronoUnit.DAYS.between(startDateFormat, dueDateFormat)
            val remainingDays = totalPlanDays - daysLeft
            val totalAmountRemainingPayed = if (remainingDays > 0) {
                val currentPriceByDay = oldPrice / totalPlanDays
                currentPriceByDay * remainingDays
            } else {
                0.0
            }
            val newPriceByDay = newPrice / totalPlanDays
            val totalAmountRemainingNotPayed = newPriceByDay * remainingDays
            totalAmountRemainingNotPayed - totalAmountRemainingPayed
        } else {
            newPrice - oldPrice
        }
    }
}